### PR TITLE
Add net481 as a netframework version

### PR DESCRIPTION
**Pull request by BenVillalobos**
Fixes ToolLocationHelper returning null when on net481. 

Fixes https://github.com/dotnet/msbuild/issues/8027 and https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1541603/

### Context
Net481 is the latest .NET Framework, but calling `Microsoft.Build.Utilities.ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.VersionLatest);` returns null. MSBuild needs to learn that net481 exists.

### Changes Made
Add entries for net481 when returning paths to various .net framework installs.

Broadens our "valid install" check to include checking for microsoft.build.dll as well as msbuild.exe. The reason for this is net481 did not include msbuild.exe because it didn't exist at the time. The binaries did ship though, so we can check for those.

### Testing
Local testing with a simple repro by replacing the Microsoft.Build.Utility.Core.dll with my custom one. The function should return the full path to al.exe.
```
            Console.WriteLine("Framework SDK File: " + ToolLocationHelper.GetPathToDotNetFrameworkSdkFile("al.exe", (TargetDotNetFrameworkVersion)15));

            Console.ReadKey();
```

### CODE REVIEW

File: src/Utilities/ToolLocationHelper.cs [hunk 1]
```diff
@@ -105,13 +105,18 @@ public enum TargetDotNetFrameworkVersion
...
+        Version481 = 15,
+
         /// <summary>
         /// The latest version available at the time of major release. This
         /// value should not be updated in minor releases as it could be a
         /// breaking change. Use 'Latest' if possible, but note the
         /// compatibility implications.
         /// </summary>
-        VersionLatest = Version48,
+        VersionLatest = Version481,
```
Comment (rainersigwald) [inline src/Utilities/ToolLocationHelper.cs:119]: The comment indicates we shouldn't change this. Do we need to? If so should the comment be updated?

File: src/Utilities/ToolLocationHelper.cs [hunk 2]
```diff
@@ -105,13 +105,18 @@ public enum TargetDotNetFrameworkVersion
         /// </summary>
         Version48 = 14,
 
+        /// <summary>
+        /// version 4.8.1
+        /// </summary>
+        Version481 = 15,
```
Comment (rainersigwald) [inline src/Utilities/ToolLocationHelper.cs:111]: Reference this around line 2052 to be the new Latest

File: src/Utilities/ToolLocationHelper.cs (hunk 1 above)
Comment (BenVillalobos) [reply src/Utilities/ToolLocationHelper.cs:119]: Presumably not. I'm trying to see what breaks in unit tests by forcing this

**Comment by BenVillalobos** [comment]
Status on this: the arm64 net481 install does not automatically come with MSBuild.exe. This results in MSBuild _almost_ finding the correct MSBuild, but not seeing MSBuild.exe there and therefore assuming it was a broken install.

https://github.com/dotnet/msbuild/blob/7a0fefe241f1040c4ebfa42626a854839fa0f71e/src/Shared/FrameworkLocationHelper.cs#L1384-L1386

You'll notice there's a `this._hasMSBuild` check, this is always set to true no matter what:
https://github.com/dotnet/msbuild/blob/7a0fefe241f1040c4ebfa42626a854839fa0f71e/src/Shared/FrameworkLocationHelper.cs#L1132-L1143

## Brain Dump Time

- Can we just ship MSBuild.exe into net481?

  - Sure, but we'd still need to account for 481 installs that don't include MSBuild.exe

- Does it really mean that the net framework install is broken if msbuild.exe doesn't exist? Clearly not.
  - Would callers of `ToolLocationHelper.GetPathToDotNetFramework` expect MSBuild.exe to exist? The API summarizes it as: `Get a fully qualified path to the frameworks root directory.`, so I don't think that's a problem.

- We can check for `Microsoft.Build.dll` instead of, or as well as MSBuild.exe
  - Is that dll always guaranteed to exist if msbuild.exe doesn't? I think it's safe to say that if we once assumed MSBuild.exe was going to exist, we can _even more safely_ assume that Microsoft.Build.dll will exist as an AnyCPU binary (correct me if I'm wrong)
    
    - Correcting myself: Previous versions within `C:/Windows/Microsoft.NET/Framework64` sometimes don't have Microsoft.Build.dll (v2.0.5, v3.5). Can we expect future versions to have this?

### Current Path Forward
If MSBuild.exe doesn't exist, check if `Microsoft.Build.dll` exists. We can handle future cases as they come up, just like this one.

**Comment by Forgind** [comment]
@BenVillalobos, your comment from June looks stale. Is this still needed? Close it?

**Comment by BenVillalobos** [comment]
After refreshing myself on this, The issue is that we're not shipping arm64 MSBuild.exe into `C:\Windows\Microsoft.NET\FrameworkArm64\v4.0.30319`.

ARM64 msbuild probably wasn't a thing when this folder was created, but it is now and we should ship msbuild.exe in there. I'm not entirely sure who to contact here. cc @marcpopMSFT 

Considering the workaround I mentioned higher up in the comments, it's not worth doing. The big problem is that msbuild isn't being shipped here.

**Comment by BenVillalobos** [comment]
After discussing with Nikola and Rainer, we decided to broaden the MSBuild.exe check.

File: src/Shared/FrameworkLocationHelper.cs [hunk 3]
```diff
@@ -1380,10 +1391,14 @@ public virtual string GetPathToDotNetFramework(DotNetFrameworkArchitecture archi
                                     Directory.GetDirectories,
                                     architecture);
 
-                // .net was improperly uninstalled: msbuild.exe isn't there
+                // Assume if either MSBuild.exe or Microsoft.Build.dll are shipped, there is a valid install.
+                // Note: net481 did not ship an MSBuild.exe, so checking its dll's is the fallback for a valid install.
```
Comment (rainersigwald) [inline src/Shared/FrameworkLocationHelper.cs:1395]:
```suggestion
                // Note: net481 did not ship an ARM64 MSBuild.exe, so checking its dll's is the fallback for a valid install.
```

File: src/Shared/FrameworkLocationHelper.cs [hunk 4]
```diff
@@ -1380,10 +1391,14 @@ public virtual string GetPathToDotNetFramework(DotNetFrameworkArchitecture archi
...
+                // Context: https://github.com/dotnet/msbuild/pull/7689
                 if (this._hasMsBuild &&
                     generatedPathToDotNetFramework != null &&
-                    !FileSystems.Default.FileExists(Path.Combine(generatedPathToDotNetFramework, NativeMethodsShared.IsWindows ? "MSBuild.exe" : "mcs.exe")))
+                    (!FileSystems.Default.FileExists(Path.Combine(generatedPathToDotNetFramework, NativeMethodsShared.IsWindows ? "MSBuild.exe" : "mcs.exe")) &&
+                     !FileSystems.Default.FileExists(Path.Combine(generatedPathToDotNetFramework, "Microsoft.Build.dll")))
```
Comment (rainersigwald) [inline src/Shared/FrameworkLocationHelper.cs:1400]: should this be ARM only?

File: src/Utilities/PublicAPI/net/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Microsoft.Build.Utilities.TargetDotNetFrameworkVersion.Version481 = 15 -> Microsoft.Build.Utilities.TargetDotNetFrameworkVersion
+Microsoft.Build.Utilities.TargetDotNetFrameworkVersion.VersionLatest = 15 -> Microsoft.Build.Utilities.TargetDotNetFrameworkVersion
```
Comment (rainersigwald) [inline src/Utilities/PublicAPI/net/PublicAPI.Unshipped.txt:2]: Looks like `VersionLatest` can go back to Shipped

File: src/Shared/FrameworkLocationHelper.cs [hunk 6]
```diff
@@ -289,6 +293,7 @@ internal static class FrameworkLocationHelper
                 dotNetFrameworkVersion471,
                 dotNetFrameworkVersion472,
                 dotNetFrameworkVersion48,
+                dotNetFrameworkVersion481,
```
Comment (rainersigwald) [inline src/Shared/FrameworkLocationHelper.cs:296]: Can you really target 4.8.1 from dev15 (this section)? Doesn't feel right to me.

File: src/Shared/FrameworkLocationHelper.cs [hunk 7]
```diff
@@ -327,6 +333,7 @@ internal static class FrameworkLocationHelper
...
```
Comment (rainersigwald) [inline src/Shared/FrameworkLocationHelper.cs:336]: This is the only one I expected, but to be super clear I'm not 100% confident I understand the current functionality of these VisualStudioSpecs

File: src/Shared/FrameworkLocationHelper.cs (hunk 4 above)
Comment (BenVillalobos) [reply src/Shared/FrameworkLocationHelper.cs:1400]: I'm indifferent about it. Non-arm scenarios will find msbuild.exe and it should short-circuit.

File: src/Shared/FrameworkLocationHelper.cs (hunk 6 above)
Comment (BenVillalobos) [reply src/Shared/FrameworkLocationHelper.cs]: Not sure why I put this here in the first place, but I agree. There was another area where I created a spec for net481 -> VS 15, which I've changed to be net481->vs17

**Comment by BenVillalobos** [comment]
> Have you validated that this works in the container scenario that was failing?

not sure about a container scenario, but I'm going to verify locally that this scenario is fixed: https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1541603/

TL;DR (for the repro): Call toollocationhelper.GetPathToDotnetFramework, expect a non-null-actual-path

**Comment by BenVillalobos** [comment]
I finally got the method to return the proper path. I need to investigate what changes in the PR are actually necessary.

cc @rainersigwald about the container scenario to check afterward.

**Comment by BenVillalobos** [comment]
Quick Recap: I figured out that the fix for the [original issue](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1541603/) just needs this one commit: 88dbe574b797b45cf0676c34a201b6b757a9934d.

I don't think the [issue Rainer mentioned](https://github.com/dotnet/msbuild/issues/8027) would be resolved by this. My current understanding is that we need another change similar to https://github.com/dotnet/msbuild/pull/4487/files. MSBuild needs to learn that 4.8.1 can exist in the registry, and apparently that needs to happen in the app.config.

**Comment by BenVillalobos** [comment]
Okay, I've validated the registry entry (after realizing this was stored in the 32 bit registry, and that the 32 bit registry is separate from the 64 bit registry /shakefist). 

It seems easy enough to update 4.8 to 4.8.1 similar to https://github.com/dotnet/msbuild/pull/4487/files. My main concern was "was if a user doesn't have 481 but they have 48?" My latest commit (235b524) captures that scenario by adding an explicit fallback rule.

**Comment by BenVillalobos** [comment]
Just verified we can find net481, and a tool under that folder (I manually replaced Microsoft.Build.Utilities in bin/debug and reran the same code without rebuilding):
image

**Comment by BenVillalobos** [comment]
Just verified this resolves https://github.com/dotnet/msbuild/issues/8027

See https://github.com/dotnet/msbuild/issues/8027#issuecomment-1312273773 for the details there.

File: src/MSBuild/app.amd64.config [hunk 8]
```diff
@@ -172,15 +172,15 @@
         <property name="MSBuildToolsPath32" value="$([MSBuild]::GetToolsDirectory32())" />
         <property name="MSBuildToolsPath64" value="$([MSBuild]::GetToolsDirectory64())" />
         <property name="MSBuildSDKsPath" value="$([MSBuild]::GetMSBuildSDKsPath())" />
-        <property name="FrameworkSDKRoot" value="$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8', 'InstallationFolder', null, RegistryView.Registry32))" />
+        <property name="FrameworkSDKRoot" value="$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8.1', 'InstallationFolder', null, RegistryView.Registry32))" />
```
Comment (rainersigwald) [inline src/MSBuild/app.amd64.config:175]: Is this safe to do? What does the default VS installer install for  the .NET SDK? Is it 4.8.1? If it's not, what happens if 4.8.1 isn't installed?

File: src/Utilities/ToolLocationHelper.cs (hunk 2 above)
Comment (rainersigwald) [reply src/Utilities/ToolLocationHelper.cs:111]: Ping on this, is it a bad idea or just missed?

File: src/Shared/FrameworkLocationHelper.cs [hunk 9]
```diff
@@ -373,6 +378,9 @@ internal static class FrameworkLocationHelper
             { (dotNetFrameworkVersion471, visualStudioVersion160), (dotNetFrameworkVersion47, visualStudioVersion160) },
             { (dotNetFrameworkVersion472, visualStudioVersion160), (dotNetFrameworkVersion471, visualStudioVersion160) },
             { (dotNetFrameworkVersion48, visualStudioVersion160), (dotNetFrameworkVersion472, visualStudioVersion160) },
+
+            // VS 17
+            { (dotNetFrameworkVersion481, visualStudioVersion170), (dotNetFrameworkVersion48, visualStudioVersion170) },
```
Comment (rainersigwald) [inline src/Shared/FrameworkLocationHelper.cs:383]: Should this be a full list of all the supported targets for VS 17? That's my naive interpretaion of the VS16 section above.

File: src/MSBuild/app.amd64.config (hunk 8 above)
Comment (BenVillalobos) [reply src/MSBuild/app.amd64.config:175]: This is what I've been most curious about. https://github.com/dotnet/msbuild/pull/4487 merged with no discussion so I don't have much context on how safe this is. At face value it looks like FrameworkSDKRoot would be empty. I'll try setting up a clean machine with a clean install and see what happens.

What we could safely do is create a "backup" property that stays on 4.8, and if the new prop is empty we fall back to the backup.

File: src/Utilities/ToolLocationHelper.cs (hunk 2 above)
Comment (BenVillalobos) [reply src/Utilities/ToolLocationHelper.cs:111]: In theory the fallback rule would save this if the user didn't have 481 installed, so we should be safe to bump VersionLatest to the next version.

File: src/Utilities/ToolLocationHelper.cs [hunk 10]
```diff
@@ -105,6 +105,11 @@ public enum TargetDotNetFrameworkVersion
...
+
         /// <summary>
         /// The latest version available at the time of major release. This
```
Comment (BenVillalobos) [inline src/Utilities/ToolLocationHelper.cs:114]: We need to bump this during the next major version.

Comment (BenVillalobos) [reply src/Utilities/ToolLocationHelper.cs:114]: Created https://github.com/dotnet/msbuild/issues/8163 to track this

File: src/Shared/FrameworkLocationHelper.cs (hunk 9 above)
Comment (BenVillalobos) [reply src/Shared/FrameworkLocationHelper.cs:383]: I was focusing a bit too much on net481 specifically here. While we're here we should include all versions compatible with vs 17

File: src/MSBuild/app.amd64.config (hunk 8 above)
Comment (rainersigwald) [reply src/MSBuild/app.amd64.config:175]: Hmm. This is the default installer (live public 17.4):

image

So the user is _not_ guaranteed to have 4.8 installed, and should be guaranteed to have 4.7.2. But I don't remember hearing any grumbling about pointing to 4.8 in the toolset. @tmeschter do you recall any additional context on #4487?

Comment (BenVillalobos) [reply src/MSBuild/app.amd64.config:175]: When removing the registry entry (in a VM 😱) I verified that calling `GetPathToDotNetFrameworkSdkFile` passing net481, it does return null.

Building a localized resource also shows this error:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		Task could not find "al.exe" using the SdkToolsPath "" or the registry key "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8\WinSDK-NetFx40Tools-x86". Make sure the SdkToolsPath is set and the tool exists in the correct processor specific location under the SdkToolsPath and that the Microsoft Windows SDK is installed	ToolLocationHelperRepro			
```

So if the key isn't there there _will_ be a build failure, which was entirely possible when we updated to the 4.8 registry key. Maybe we wait until the next major version for that update as well?

Comment (BenVillalobos) [reply src/MSBuild/app.amd64.config:175]: ^ this is also an argument for having the "backup" property where we query for 4.8.1 and 4.8, and if 4.8.1 is empty then we use 4.8

**Comment by BenVillalobos** [comment]
I added the fallback logic if the net481 registry key doesn't exist. It should fall back to net48 (the last version we knew was okay) so we can avoid regressing customers for now.

File: src/MSBuild/app.amd64.config [hunk 11]
```diff
@@ -181,6 +181,7 @@
         <property name="MSBuildFrameworkToolsRoot" value="$(SystemRoot)\Microsoft.NET\Framework\" />
         <property name="SDK35ToolsPath" value="$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v8.0A\WinSDK-NetFx35Tools-x86', 'InstallationFolder', null, RegistryView.Registry32))" />
         <property name="SDK40ToolsPath" value="$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8\WinSDK-NetFx40Tools-x86', 'InstallationFolder', null, RegistryView.Registry32))" />
+        <property name="SDK481ToolsPath" value="$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8.1\WinSDK-NetFx40Tools-x86', 'InstallationFolder', null, RegistryView.Registry32))" />
```
Comment (rainersigwald) [inline src/MSBuild/app.amd64.config:184]: Another option would avoid the need for downstream changes:

```suggestion
        <property name="SDK40ToolsPath" value="$([MSBuild]::ValueOrDefault($([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8\WinSDK-NetFx40Tools-x86', 'InstallationFolder', null, RegistryView.Registry32)), $([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8.1\WinSDK-NetFx40Tools-x86', 'InstallationFolder', null, RegistryView.Registry32))" />
```

Think that's better, or too complicated to be helpful?

**Review by rainersigwald — COMMENTED** [review]
> My last compatibility question here is: Should we add versions for net481 on VS 16,15, etc.? How far back do we go?

I think leave it in only 17.

File: src/MSBuild/app.amd64.config (hunk 11 above)
Comment (BenVillalobos) [reply src/MSBuild/app.amd64.config:184]: This works, though I think it should be ValueOrDefault(<481 registry key>, <48 registry key>) so we can fall back to 48

Comment (rainersigwald) [reply src/MSBuild/app.amd64.config:184]: Yes, definitely!

Comment (BenVillalobos) [reply src/MSBuild/app.amd64.config:184]: I verified this logic works and correctly falls back to 4.8 after I uninstalled 4.8.1.

(see "Foo: Test: ...")
```
C:\src\temp\11-15\9_36>msbuild
MSBuild version 17.5.0-preview-22519-01+82a30d307 for .NET Framework
Build started 11/15/2022 9:38:25 AM.
Project "C:\src\temp\11-15\9_36\proj.proj" on node 1 (default targets).
Foo:
 <SEE THIS>---->  Test: C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8.1 Tools\
...
[C:\src\temp\11-15\9_36\proj.proj]
CopyFilesToOutputDirectory:
  proj ->
Done Building Project "C:\src\temp\11-15\9_36\proj.proj" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.44

C:\src\temp\11-15\9_36>msbuild
MSBuild version 17.5.0-preview-22519-01+82a30d307 for .NET Framework
Build started 11/15/2022 9:42:53 AM.
Project "C:\src\temp\11-15\9_36\proj.proj" on node 1 (default targets).
Foo:
<SEE THIS>---->  Test: C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\
...
 [C:\src\temp\11-15\9_36\proj.proj]
CopyFilesToOutputDirectory:
  proj ->
Done Building Project "C:\src\temp\11-15\9_36\proj.proj" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.44
```

The project:
```xml
<Project Sdk="Microsoft.Build.NoTargets/3.1.0" InitialTargets="Foo">

    <PropertyGroup>
        <TargetFramework>net472</TargetFramework>
        <Test>$([MSBuild]::ValueOrDefault($([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8.1\WinSDK-NetFx40Tools-x86', 'InstallationFolder', null, RegistryView.Registry32)), $([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\NETFXSDK\4.8\WinSDK-NetFx40Tools-x86', 'InstallationFolder', null, RegistryView.Registry32))))</Test>
    </PropertyGroup>

    <Target Name="Foo">
        <Message Text="Test: $(Test)" Importance="High" />
    </Target>

</Project>

```

**Review by rainersigwald — APPROVED** [review]
